pkg/scheduler/queue/queue.go

```diff
@@ -54,229 +54,354 @@ type Request interface{}
+	q.querierOperations = make(chan querierOperation)
+	q.enqueueRequests = make(chan enqueueRequest)
+	q.availableQueriers = make(chan *availableQuerier)
+
+	go q.dispatcherLoop()
+
+	return nil
+}
+
+func (q *RequestQueue) dispatcherLoop() {
+	stopping := false
+	queues := newUserQueues(q.maxOutstandingPerTenant, q.forgetDelay)
+
+	var firstAvailableQuerier *availableQuerier
+	var lastAvailableQuerier *availableQuerier
+
+	for {
+		needToDispatchQueries := false
+
+		select {
+		case <-q.stopRequested:
+			// Nothing much to do here - fall through to the stop logic below to see if we can stop immediately.
+			stopping = true
+		case qe := <-q.querierOperations:
+			switch qe.operation {
+			case registerConnection:
+				queues.addQuerierConnection(qe.querierID)
+			case unregisterConnection:
+				queues.removeQuerierConnection(qe.querierID, time.Now())
+			case notifyShutdown:
+				queues.notifyQuerierShutdown(qe.querierID)
+			case forgetDisconnected:
+				if queues.forgetDisconnectedQueriers(time.Now()) > 0 {
+					// Removing some queriers may have caused a resharding.
+					needToDispatchQueries = true
+				}
+			default:
+				panic(fmt.Sprintf("received unknown querier event %v for querier ID %v", qe.operation, qe.querierID))
+			}
+			qe.processed <- struct{}{}
+		case r := <-q.enqueueRequests:
+			err := q.handleNewRequest(queues, r)
+			r.processed <- err
+
+			if err == nil {
+				// TODO: might be able to be much smarter here and try to find a querier that can take the request directly
+				needToDispatchQueries = true
+			}
+		case querier := <-q.availableQueriers:
+			if !q.dispatchRequestToQuerier(queues, querier) {
+				// No requests available for this querier right now. Add it to the list to try later.
+				if lastAvailableQuerier == nil {
+					firstAvailableQuerier = querier
+					lastAvailableQuerier = querier
+				} else {
+					lastAvailableQuerier.next = querier
+					lastAvailableQuerier = querier
+				}
+			}
+		}
+
+		if needToDispatchQueries {
+			var previousQuerier *availableQuerier
+			currentQuerier := firstAvailableQuerier
+
+			for currentQuerier != nil && queues.len() > 0 {
+				if q.dispatchRequestToQuerier(queues, currentQuerier) {
+					// Remove this querier from our list.
+
+					if previousQuerier == nil {
+						// This was the first available querier.
+						firstAvailableQuerier = currentQuerier.next
+					} else {
+						previousQuerier.next = currentQuerier.next
+					}
+
+					if currentQuerier.next == nil {
+						// This was the last available querier.
+						lastAvailableQuerier = previousQuerier
+					}
+
+					currentQuerier = currentQuerier.next
+				} else {
+					// Nothing for this querier right now, advance through the list.
+					previousQuerier = currentQuerier
+					currentQuerier = currentQuerier.next
+				}
+			}
+		}
+
+		if stopping && (queues.len() == 0 || q.connectedQuerierWorkers.Load() == 0) {
```

Comment:
if `connectedQuerierWorkers == 0` but there are active queries in the queue, then we'd abandom them, is this right?